src/python/zquantum/core/evolution.py

```diff
@@ -1,157 +1,250 @@
+    if isinstance(hamiltonian, QubitOperator):
+        terms = list(hamiltonian.get_operators())
+    elif isinstance(hamiltonian, pyquil.paulis.PauliSum):
+        warnings.warn(
+            "PauliSum as an input to time_evolution will be depreciated, please change to QubitOperator instead.",
+            DeprecationWarning,
+        )
+        terms = hamiltonian.terms
+
+    return reduce(
+        operator.add,
+        (
+            time_evolution_for_term(term, time / trotter_order)
+            for _index_order in range(trotter_order)
+            for term in terms
+        ),
+    )
+
+
+@singledispatch
+def time_evolution_for_term(term, time: Union[float, sympy.Expr]):
+    raise NotImplementedError
+
+
+@time_evolution_for_term.register
+def _time_evolution_for_term_pyquil(
+    term: pyquil.paulis.PauliTerm, time: Union[float, sympy.Expr]
+) -> Circuit:
+    """Evolves a Pauli term for a given time and returns a circuit representing it.
+    Args:
+        term: Pauli term to be evolved
+        time: time of evolution
+    Returns:
+        Circuit: Circuit representing evolved pyquil term.
+    """
+    if isinstance(time, sympy.Expr):
+        circuit = Circuit(pyquil.paulis.exponentiate(term))
+        for gate in circuit.gates:
+            if len(gate.params) > 1:
+                raise (
+                    NotImplementedError(
+                        "Time evolution of multi-parametered gates with symbolic "
+                        "parameters is not supported."
+                    )
                 )
-
+            elif gate.name == "Rz" or gate.name == "PHASE":
+                # We only want to modify the parameter of Rz gate or PHASE gate.
+                gate.params[0] = gate.params[0] * time
     else:
-        raise ValueError("Currently the method {} is not supported".format(method))
+        exponent = term * time
+        assert isinstance(exponent, pyquil.paulis.PauliTerm)
+        circuit = Circuit(pyquil.paulis.exponentiate(exponent))
+    return circuit
+
+
+@time_evolution_for_term.register
+def _time_evolution_for_term_qubit_operator(
+    term: QubitOperator, time: Union[float, sympy.Expr]
+) -> Circuit:
+    """Evolves a Pauli term for a given time and returns a circuit representing it.
+    Args:
+        term: Pauli term to be evolved
+        time: time of evolution
+    Returns:
+        Circuit: Circuit representing evolved pyquil term.
+    """
+
+    if len(term.terms) != 1:
+        raise ValueError("This function works only on a single term.")
+    term_components = list(term.terms.keys())[0]
+    base_changes = []
+    base_reversals = []
+    cnot_gates = []
+    central_gate = None
+    term_types = [component[1] for component in term_components]
+    qubit_indices = [component[0] for component in term_components]
+    coefficient = list(term.terms.values())[0]
+
+    for i, (term_type, qubit_id) in enumerate(zip(term_types, qubit_indices)):
+        # TODO: comments
```

Comment:
Maybe instead of leaving TODO here, we could include a link to some arxiv describing algorithm implemented here?